### PR TITLE
pthread/spinlock: Call up_testsest directly in the flat build 

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary

- libc: Move the common implementation of up_testset to libc/machine 

## Impact
spinlock

## Testing
Pass CI

### CODE REVIEW

**Review by Ouss4 — COMMENTED** [review]
Change looks good.
I'm just wondering if `arch_atomic.c` is the best place for the `up_testset` function.

cc. @gustavonihei

File: include/nuttx/spinlock.h [hunk 1]
```diff
@@ -107,28 +107,7 @@ typedef struct
  *
  ****************************************************************************/
 
-#if defined(CONFIG_ARCH_HAVE_TESTSET)
 spinlock_t up_testset(volatile FAR spinlock_t *lock);
-#elif !defined(CONFIG_SMP)
```
Comment (gustavonihei) [inline include/nuttx/spinlock.h:112]: There is a change in the logic here, now this function will be built for SMP targets. Is this really intended?

Comment (Ouss4) [reply include/nuttx/spinlock.h:112]: I don't think so.  The condition is `#if defined(CONFIG_SPINLOCK) && !defined(CONFIG_ARCH_HAVE_TESTSET)` and SMP depends on `CONFIG_ARCH_HAVE_TESTSET`.

Comment (gustavonihei) [reply include/nuttx/spinlock.h:112]: You’re right, thanks for the clarification @Ouss4!

**Comment by xiaoxiang781216** [comment]
> Change looks good. I'm just wondering if `arch_atomic.c` is the best place for the `up_testset` function.
> 
> cc. @gustavonihei

up_testset is one of atomic function. Actually, it isn't good to introduce up_testset since atomic is part of C/C++ standard. It's better to use the standard atomic api, and implement them on the old toolchain by hand.

**Comment by gustavonihei** [comment]
> > Change looks good. I'm just wondering if `arch_atomic.c` is the best place for the `up_testset` function.
> > cc. @gustavonihei
> 
> up_testset is one of atomic function.

The part that makes this new location a bit weird is that `arch_atomic.c` is under `libc`, and `up_testset` is specific for spinlocks.

> Actually, it isn't good to introduce up_testset since atomic is part of C/C++ standard. It's better to use the standard atomic api, > and implement them on the old toolchain by hand.

So another option would be to leave `up_testset` in its place and change it to use the C11 function. And then on `arch_atomic.c`  we could provide the implementation for the GCC/Clang builtins for those archs that don't support.

**Comment by xiaoxiang781216** [comment]
Ok, I drop up_testset from PR, let's create a new PR to remove up_testset and call atomic api instead.

**Comment by xiaoxiang781216** [comment]
Here is https://github.com/apache/incubator-nuttx/pull/6283

**Review by gustavonihei — APPROVED** [review]
LGTM

**Review by pkarashchenko — APPROVED** [review]

**Comment by Ouss4** [comment]
> Here is #6283

Recursion. :)

**Comment by xiaoxiang781216** [comment]
should be https://github.com/apache/incubator-nuttx/issues/6301 :)

**Pull request merged**